keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,250 @@
+
+This proposal introduces three stability classes for metrics: (1) Alpha, (2) Stable, (3) Deprecated. These classes are intended to make explicit the API contract between the control-plane and the ingester of control-plane metrics.
+
+__Alpha__ metrics have __*no*__ stability guarantees; as such they can be modified or deleted at any time. At this time, all kubernetes metrics implicitly fall into this category.
+
+__Stable__ metrics can be guaranteed to *not change*, except that the metric may become marked deprecated for a future kubernetes version. By *not change*, we mean three things:
+
+1. the metric itself will not be deleted
+2. the type of metric will not be modified
+3. no labels can be added or removed from this metric
+
+From an ingestion point of view, it is backwards-compatible to add or remove possible __values__ for labels which already do exist (but __not__ labels themselves). Therefore, adding or removing __values__ from an existing label is permissible. Stable metrics can also be marked as __deprecated__ for a future kubernetes version, since this is a metadata field and does not actually change the metric itself.
+
+__Deprecated__ metrics are also guaranteed to *not change*. The purpose of deprecated metrics is to provide a reasonable window for consumers of control-plane metrics to make changes to their monitoring infrastructure (and also so that we do not handcuff control-plane developers to a set of stable metrics which they will have to then support for all time). Deprecated metrics can fall under one of two categories:
+
+ 1. _soft deprecated_ - this occurs when a stable metric has been recently deprecated; 'recently deprecated' is defined as when the metric's deprecated version is equal to the current kubernetes version
+ 2. _hard deprecated_ - this occurs when the metric has been deprecated for > 1 release.
+
+Soft deprecated metrics will have their description text prefixed with a deprecation notice string '(Deprecated from x.y)' and a warning log will be emitted during metric registration (in the spirit of the official [kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli)). Like their stable metric counterparts, soft deprecated metrics will be automatically registered to the metrics endpoint.
+
+Metrics are _hard deprecated_ once they have been deprecated for more than one release. _Unlike_ their soft deprecated counterparts, hard deprecated metrics will __*no longer be automatically registered*__ to the metrics endpoint. By default, hard deprecated metrics will not be registered unless explicitly instructed to do so through command line flag on the binary (i.e. '--enable-hard-deprecated-metrics=all-metrics'). It is expected that deprecated metrics will be removed once they have been deprecated for more than one release. As such, metrics which have been deprecated for more than one release will have __no__ stability guarantees.
```

Comment:
This is confusing.  The lifecycle is that we can hard deprecate it, but sometimes maybe it might not get deleted.  But it might.  But if it's not you could turn it on with flag?

I would prefer that we simplify this and say "they are deprecated for one release, then delted".

Reply:
I wanted to keep the two cycle deprecation policy so that we have one release cycle where the metric works and we attempt to warn the cluster admins that the metric is deprecated, and one release cycle where something would break but the cluster admin would have an escape hatch in case they required some time to get changes into their monitoring backend. The additional logic would be pretty trivial to add to the custom metrics registry but would be considerably friendlier, policy-wise, for cluster admins.

Reply:
So the steps are stable -> soft deprecated -> hard deprecated -> gone?  I don't think that's quite made clear here.  You say "It is expected that [...]", but why don't we just say "Hard deprecated metrics will be removed the next cycle."?

Reply:
Ah.. I see, I will make it more clear what I mean.

Reply:
Updated with the suggestions. Thanks for taking the cycles on this!